### PR TITLE
[MLDB-1468] removed credential daemon

### PR DESCRIPTION
**Pull request by guyd**
- removed references to the credential daemon
- integrated the creation of credentials in MLDB

I am splitting this into two PR to facilitate the review.  This is mostly removal of the references to the credential daemon and an update of the documentations.

There is more logic to change but I am afraid that adding it here will simply make it harder to review.  What is remaining:
- fix the logic to lookup credentials
- complete the test coverage on credential collection once this is done

### CODE REVIEW

File: container_files/public_html/doc/builtin/Url.md [hunk 1]
```diff
@@ -46,14 +46,12 @@ external tools.
...
 ## Credentials
 
-MLDB uses an subprogram known as `credentialsd` to keep credentials and supply them to the main
-MLDB process when required.
+MLDB can store credentials and supply them whenever required.
 
-This allows for MLDB users to control how credentials are stored and used, and
-to create temporary credentials for MLDB to limit its permissions to exactly what are needed.
+Capabilities have been added to control how credentials are stored and used, and
+to create temporary credentials to limit its permissions to exactly what are needed.
```
Comment (jeremybarnes) [inline container_files/public_html/doc/builtin/Url.md]: This isn't true anymore.  With credentialsd, you could point MLDB to another external daemon which did this.  Now we no longer can, so we should remove this from the doc.

File: container_files/public_html/doc/builtin/Url.md [hunk 2]
```diff
@@ -63,15 +61,17 @@ Credentials are specified by giving rules.  A credentials rule contains two part
...
-in which the request is being made.  The credenials service will scan through its list
-of rules, and return the credentials of the first matching rule to satisfy the request.
+When a resource that requires credentials is requested, MLDB will scan the stored
+rules for those credentials, passing it the resource and the context (user id, etc)
+in which the request is being made.  MLDB will use the first matching rule that satisfies
+the request.
 
 ### Credentials storage
 
-Credentials are stored in the `mldb_data` directory under the `.mldb_credentials` subdirectory.
```
Comment (jeremybarnes) [inline container_files/public_html/doc/builtin/Url.md]: Where are the credentials stored on disk?  A user may want to know in order to understand what is stored, debug, or remove credentials.

File: server/credential_collection.cc [hunk 3]
```diff
@@ -167,16 +221,45 @@ getConfig(std::string key, const CredentialRule & value) const
...
+    CredentialProvider::registerProvider("collectionCredentials",
+                                         std::make_shared<CollectionCredentialProvider>(result));
+    return result;
+}
 
 /*****************************************************************************/
 /* CREDENTIAL DAEMON                                                         */
 /*****************************************************************************/
-
+#if 0
```
Comment (jeremybarnes) [inline server/credential_collection.cc]: Delete?

File: server/credential_collection.h [hunk 4]
```diff
@@ -91,16 +90,11 @@ struct CredentialRuleCollection
     getConfig(std::string key, const CredentialRule & value) const;
 };
 
-extern template class RestCollection<std::string, CredentialRule>;
```
Comment (jeremybarnes) [inline server/credential_collection.h]: Why remove this?  It makes the compilation faster.

File: server/credential_collection.h [hunk 5]
```diff
@@ -91,16 +90,11 @@ struct CredentialRuleCollection
...
-extern template class RestConfigurableCollection<std::string, CredentialRule,
-                                                 CredentialRuleConfig,
-                                                 CredentialRuleStatus>;
-
 
+#if 0
```
Comment (jeremybarnes) [inline server/credential_collection.h]: Remove?

File: server/dataset_collection.cc [hunk 6]
```diff
@@ -26,12 +26,10 @@ namespace Datacratic {
...
 
 std::shared_ptr<DatasetCollection>
-createDatasetCollection(MldbServer * server, RestRouteManager & routeManager,
-                       std::shared_ptr<CollectionConfigStore> configStore)
+createDatasetCollection(MldbServer * server, RestRouteManager & routeManager)
 {
     return createCollection<DatasetCollection>(2, L"dataset", L"datasets",
-                                               server, routeManager,
-                                               configStore);
+                                               server, routeManager, nullptr);
```
Comment (jeremybarnes) [inline server/dataset_collection.cc]: Maybe make nullptr the default and simply remove the parameter?

File: server/mldb_server.cc [hunk 7]
```diff
@@ -97,27 +98,9 @@ MldbServer::
     shutdown();
 }
 
-#if 0
```
Comment (jeremybarnes) [inline server/mldb_server.cc]: Either re-enable or remove?

File: soa/service/remote_credential_provider.cc [hunk 8]
```diff
@@ -1,188 +0,0 @@
-// This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.
```
Comment (jeremybarnes) [inline soa/service/remote_credential_provider.cc]: Consider keeping this.  It provides an escape hatch for if the user doesn't want to trust internal credentials, and is orthogonal to the rest of the changes.

File: soa/service/remote_credential_provider.h [hunk 9]
```diff
@@ -1,93 +0,0 @@
-// This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.
```
Comment (jeremybarnes) [inline soa/service/remote_credential_provider.h]: Ditto

File: testing/credentials_daemon_test.cc [hunk 10]
```diff
@@ -1,344 +0,0 @@
-/** credentials_daemon_test.cc
```
Comment (jeremybarnes) [inline testing/credentials_daemon_test.cc]: Can we keep this test program, but make it start MLDB instead of the daemon and test the MLDB credentials functionality?

File: testing/credentials_persistence_test.cc [hunk 11]
```diff
@@ -0,0 +1,271 @@
+/** credentials_persistence_test.cc
```
Comment (jeremybarnes) [inline testing/credentials_persistence_test.cc]: Ah, I see you did it already.

**Comment by jeremybarnes** [comment]
There are several comments; +1 once you've considered them.

File: server/credential_collection.cc (hunk 3 above)
Comment (guyd) [inline server/credential_collection.cc]: I will.  It was to minimize the diff size which was already too large.

File: server/mldb_server.cc (hunk 7 above)
Comment (guyd) [inline server/mldb_server.cc]: I removed it.  This is why it is part of the diff!

File: soa/service/remote_credential_provider.cc (hunk 8 above)
Comment (guyd) [inline soa/service/remote_credential_provider.cc]: As it is, users have no ways to configure an external credential provider.  If you feel that this is required than we would need to add a way to configure it and document clearly what the REST interface of that provider is.  My feeling is that if someone has this level of expertise he may as well modify the MLDB code to get what he wants.  Any strong feeling here?

Comment (nicolaskruchten) [inline soa/service/remote_credential_provider.cc]: My feeling is that this is the type of complexity we should remove for now, even if we end up adding it back in later. I think we have many hurdles to clear before "lack of external creds provider" is an objection that stops any uptake.

File: server/mldb_server.cc (hunk 7 above)
Comment (jeremybarnes) [inline server/mldb_server.cc]: ahem.  yes!

File: soa/service/remote_credential_provider.cc (hunk 8 above)
Comment (jeremybarnes) [inline soa/service/remote_credential_provider.cc]: OK, that's fine.  It was designed as something that is entirely orthogonal: you add an environment variable, you get something that will be called for credentials and it doesn't affect any other functionality nor is it enabled by defuault.  Like I said, escape hatch.

File: container_files/public_html/doc/builtin/Url.md (hunk 2 above)
Comment (marccardinal) [inline container_files/public_html/doc/builtin/Url.md]: Might want to add if they are stored in clear text

File: testing/MLDB-1003-s3-load-dataset.js [hunk 12]
```diff
@@ -17,7 +17,7 @@ function assertEqual(expr, val, msg)
 csv_conf = {
     type: "import.text",
     params: {
-        dataFileUrl : "s3://benchm-ml--main/test.csv",
+        dataFileUrl : "s3://dev.mldb.datacratic.com/testing/MLDB-1003/test.csv",
```
Comment (marccardinal) [inline testing/MLDB-1003-s3-load-dataset.js]: So everyone using the open-source mldb need access to that bucket?

Comment (guyd) [inline testing/MLDB-1003-s3-load-dataset.js]: Actually, there is something strange here.  I cannot access that public bucket s3://benchm-ml--main/test.csv (Nicolas neither) with my credentials but jenkins can.  My solution was to move the file to the dev.mldb.datacratic.com bucket but if we can fix it differently that would be great.  Any ideas?

Regarding accessing private resources in an open-source project, I've created MLDBFB-484 earlier this week because this concerns me as well.

Comment (guyd) [inline testing/MLDB-1003-s3-load-dataset.js]: OK.  I know what is happening.  Because the test fail once with my credentials and that I could not list the bucket, I concluded the I could not access that resource.  Now that I know better, thanks to Wolfgang, and understand that listing and accessing are two different beasts, I try the test and it worked.  I will remove that change.  MLDBFB-484 still remain open because there is at least one more public test that relies on private resources.

Comment (marccardinal) [inline testing/MLDB-1003-s3-load-dataset.js]: @nicolaskruchten should the files be moved to public.mldb.ai ? dev.mldb.datacratic.com sounds so internal...

Comment (guyd) [inline testing/MLDB-1003-s3-load-dataset.js]: With my last commits we are pointing to the s3://benchm-ml--main/test.csv.  Basically, I undid that change.

**Comment by nicolaskruchten** [comment]
Any outstanding issues blocking this PR?

**Comment by guyd** [comment]
This PR is waiting on its sister PR 318 to get the final go.  They go together.
